File: figures/08-graphs/snippet-kruskal.java
    public static Result kruskal(Graph g) {
        requireUndirected(g);
        List<MstEdge> edges = new ArrayList<>();
        for (int u = 0; u < g.vertices(); u++)
            for (Graph.Edge e : g.edgesFrom(u)) if (u < e.to())
                edges.add(new MstEdge(u, e.to(), e.weight()));
        edges.sort(Comparator.comparingLong(MstEdge::weight)
            .thenComparingInt(MstEdge::u).thenComparingInt(MstEdge::v));
        DisjointSet dsu = new DisjointSet(g.vertices());
        List<MstEdge> out = new ArrayList<>(); long total = 0;
        for (MstEdge e : edges) if (dsu.union(e.u(), e.v())) {
            out.add(e); total += e.weight();
            if (out.size() == g.vertices() - 1) break;
        }
        return new Result(List.copyOf(out), total, out.size() == Math.max(0, g.vertices() - 1));
    }

File: figures/08-graphs/snippet-kruskal.py
def kruskal(g):
    edges = sorted(
        ((u, v, w) for u in range(g.vertices()) for v, w in g.edges_from(u) if u < v),
        key=lambda e: (e[2], e[0], e[1]),
    )
    dsu = DisjointSet(g.vertices())
    selected = []
    total = 0
    for u, v, w in edges:
        if dsu.union(u, v):
            selected.append((u, v, w))
            total += w
            if len(selected) == g.vertices() - 1:
                break
    return selected, total
